lib/db/tag-type-store.js

```diff
@@ -0,0 +1,135 @@
+'use strict';
+
+const metricsHelper = require('../metrics-helper');
+const {
+    CREATE_TAG_TYPE,
+    DELETE_TAG_TYPE,
+    UPDATE_TAG_TYPE,
+} = require('../command-type');
+const { TAG_TYPE_CREATED, TAG_TYPE_DELETED } = require('../event-type');
+const { DB_TIME } = require('../events');
+const NotFoundError = require('../error/notfound-error');
+
+const COLUMNS = ['name', 'description', 'icon'];
+const TABLE = 'tag_types';
+
+class TagTypeStore {
+    constructor(db, eventStore, eventBus, getLogger) {
+        this.db = db;
+        this.eventStore = eventStore;
+        this.logger = getLogger('tag-type-store.js');
+        this.timer = action =>
+            metricsHelper.wrapTimer(eventBus, DB_TIME, {
+                store: 'tag-type',
+                action,
+            });
+
+        eventStore.on(CREATE_TAG_TYPE, event =>
+            this._createTagType(event.data),
+        );
+        eventStore.on(DELETE_TAG_TYPE, event =>
+            this._deleteTagType(event.data),
+        );
+        eventStore.on(UPDATE_TAG_TYPE, event =>
+            this._updateTagType(event.data),
+        );
+    }
+
+    async getAll() {
+        const stopTimer = this.timer('getTagTypes');
+        const rows = await this.db.select(COLUMNS).from(TABLE);
+        stopTimer();
+        return rows.map(this.rowToTagType);
+    }
+
+    async getTagType(query) {
+        const stopTimer = this.timer('getTagTypeByName');
+        const tagType = this.db
+            .first(COLUMNS)
+            .from(TABLE)
+            .where(query)
+            .then(row => {
+                if (!row) {
+                    throw new NotFoundError('Could not find tag-type');
+                } else {
+                    return this.rowToTagType(row);
+                }
+            });
+        stopTimer();
+        return tagType;
+    }
+
+    async _createTagType(event) {
+        const stopTimer = this.timer('createTagType');
+        try {
+            const data = this.eventDataToRow(event);
+            await this.db(TABLE)
+                .insert(data)
+                .onConflict('name')
+                .merge({
+                    description: data.description,
+                });
+            await this.eventStore.store({
```

Comment:
Not sure why you have this call here? Is not `_ceateTagType` called as a result of event-store storing this, and this emitting an event which this store is listening to? I think this will lead to a infinite loop...

Reply:
Different event, CREATE_TAG_TYPE vs TAG_TYPE_CREATED. So command vs event (both on the event store...). Since nothing is setup to react on TAG_TYPE_CREATED yet this should go fine, but it does go to show that our stores are doing to much atm.